src/vision/status_quo/barbara_wishes_for_easy_runtime_switch.md

```diff
@@ -0,0 +1,68 @@
+# 😱 Status quo stories: Barbara wishes for an easy runtime switch
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from
+real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async
+Rust programmers face today.
+
+## The story
+
+[Barbara] has been working on an async codebase for the past 5 years. It is extremely mature and
+quite large (in the millions of lines of code). The product has traditionally been run on the server
+only, but now they want to move large parts to the edge. After some investigation, they realize that
+the open-source async runtime they have been using (tokio), while great for server workloads, is
+not very performant on the low-end edge devices.
+
+Tokio makes tradeoffs in memory for performance and while performance is still a concern on
+the low-end edge devices Barbara is targeting, memory consumption is an even harder constraint there.
+Additionally, the usage of a thread pool by tokio does not help either as the targeted edge
+devices have either a single CPU thread or only a few.
+
+They had never planned to run on edge devices and so it was assumed that they would always stick
+to tokio. Barbara looks into the how much work it would be to convert their codebase from
+tokio to a custom runtime tuned to their workload. To their disappointment, they discover many
+areas, where their choice of runtime permeates the code base:
+
+* tokio provides variants of helpers macros and types, like `tokio::select!` and `tokio::Mutex`. These helpers can be used without the rest of tokio, and there are also alternatives from the futures crate and elsewhere (albeit with subtle differences).
+    * tokio uses a custom version of `AsyncRead` and `AsyncWrite` traits which differ from the ones used by other parts of the ecosystem.
+* The tokio API is needed to create core runtime operations like timers (`tokio::time::sleep`) and to launch tasks; there doesn't seem to be a standard way to abstract over those kinds of things in a runtime-independent way.
+* Some of their dependencies (e.g `hyper` and `reqwest`) are tied to tokio. In some cases, there are configuration options or ways to use those dependencies that don't depend on tokio, but there is no standard mechanism for that.
```

Comment:
Just wanted to add my two cents here. Afaik `reqwest`, although built on top `hyper` doesn't expose the `Executor` trait. But `hyper` is runtime agnostic, and you could write your own connector for it, and make a client with a custom executor.

Reply:
@emmanuelantony2000 Thanks, they're just random examples though. Are there better ones that **are** tied to runtimes?

Reply:
> Just wanted to add my two cents here. Afaik `reqwest`, although built on top `hyper` doesn't expose the `Executor` trait.

Not sure that matters here. If they've a hard dep on `hyper` and `hyper` has a hard dep on tokio, this is not ideal at all.

> But `hyper` is runtime agnostic, and you could write your own connector for it, and make a client with a custom executor.

From what I understand from the docs, it provides an agnostic API but it itself still relies on tokio unconditionally. If the task at hand is to switch to another runtime, this is a problem.

Reply:
> @emmanuelantony2000 Thanks, they're just random examples though. Are there better ones that **are** tied to runtimes?

I think `reqwest` is a good example as it is strongly tied to tokio. I don't think it exposes an executor API, like hyper.

> Not sure that matters here. If they've a hard dep on `hyper` and `hyper` has a hard dep on tokio, this is not ideal at all.
> 
> From what I understand from the docs, it provides an agnostic API but it itself still relies on tokio unconditionally. If the task at hand is to switch to another runtime, this is a problem.

It is not an easy switch. But if I'm correct, hyper just depends on the AsyncWrite and AsyncRead traits in tokio, not the runtime itself.

Probably it could be something like "hyper although runtime agnostic, required you to write a Connector if you want to use a custom runtime."

Reply:
> Probably it could be something like "hyper although runtime agnostic, required you to write a Connector if you want to use a custom runtime."

The problem is that even if it'd make it super easy to use other runtimes, it still depends on tokio itself and if the goal is to switch to another runtime, it's not nice to still depend on tokio indirectly and sort of beats the point of the exercise.